hbase-server/src/main/java/org/apache/hadoop/hbase/master/MobFileCompactionChore.java

```diff
@@ -0,0 +1,224 @@
+      LOG.error("Failed to compact", e);
+    }
+  }
+
+  @VisibleForTesting
+  public void performMajorCompactionInBatches(Admin admin, TableDescriptor htd,
+      ColumnFamilyDescriptor hcd) throws IOException, InterruptedException {
+
+    List<RegionInfo> regions = admin.getRegions(htd.getTableName());
+    if (regions.size() <= this.regionBatchSize) {
+      LOG.debug("Performing compaction in non-batched mode, regions={}, batch size={}",
+        regions.size(), regionBatchSize);
+      admin.majorCompact(htd.getTableName(), hcd.getName());
+      return;
+    }
+    // Shuffle list of regions in case if they come ordered by region server
+    Collections.shuffle(regions);
+    // Create first batch
+    List<RegionInfo> toCompact = new ArrayList<RegionInfo>();
+    for (int i = 0; i < this.regionBatchSize; i++) {
+      toCompact.add(regions.remove(0));
+    }
+
+    // Start compaction now
+    for (RegionInfo ri : toCompact) {
+      startCompaction(admin, htd.getTableName(), ri, hcd.getName());
+    }
+
+    List<RegionInfo> compacted = new ArrayList<RegionInfo>();
+    int totalCompacted = 0;
+    while (!toCompact.isEmpty()) {
+      // Check status of active compactions
+      for (RegionInfo ri : toCompact) {
+        try {
+          if (admin.getCompactionStateForRegion(ri.getRegionName()) == CompactionState.NONE) {
+            totalCompacted++;
+            LOG.info("Finished major compaction: table={} region={}, compacted regions={}",
+              htd.getTableName(), ri.getRegionNameAsString(), totalCompacted);
+            compacted.add(ri);
+          }
+        } catch (IOException e) {
+          LOG.warn("Could not get compaction state for region {}", ri.getEncodedName());
+        }
+      }
+      // Update batch: remove compacted regions and add new ones
+      for (RegionInfo ri : compacted) {
+        toCompact.remove(ri);
+        if (regions.size() > 0) {
+          RegionInfo region = regions.remove(0);
+          startCompaction(admin, htd.getTableName(), region, hcd.getName());
+          toCompact.add(region);
+        }
+      }
+      compacted.clear();
+      LOG.debug("Wait for 10 sec, toCompact size={} regions left={} compacted so far={}",
+        toCompact.size(), regions.size(), totalCompacted);
+      Thread.sleep(10000);
+    }
+    LOG.info("Finished major compacting {}. cf={}", htd.getTableName(), hcd.getNameAsString());
+
+  }
+
+  private void startCompaction(Admin admin, TableName table, RegionInfo region, byte[] cf)
+      throws IOException, InterruptedException {
+
+    LOG.info("Started major compaction: table={} region={}", table, region.getRegionNameAsString());
+    admin.majorCompactRegion(region.getRegionName());
+    // Wait until it really starts
+    // but with finite timeout
+    long waitTime = 300000; // 5 min
+    long startTime = EnvironmentEdgeManager.currentTime();
+    while (admin.getCompactionStateForRegion(region.getRegionName()) == CompactionState.NONE) {
```

Comment:
we end up being racy if there was already a compaction happening for this region right? like we're skip right through here and then when the already-in-progress compaction finishes we'll treat that as the end of the compaction we requested?

probably an okay failure mode, but include a comment about that for whoever is reading through here later?

Reply:
I do not see any race conditions here. If compaction is in progress we break the loop immediately, if - not we wait up to 5min then break the loop.